CSSParserFastPaths should treat rgb() and rgba() as synonyms

roberto-apple opened the pull request:
#### 1f96e1bf6bd0a4a1ffa84b01c50aa5639a357c9b
<pre>
CSSParserFastPaths should treat rgb() and rgba() as synonyms
<a href="https://bugs.webkit.org/show_bug.cgi?id=276761">https://bugs.webkit.org/show_bug.cgi?id=276761</a>
<a href="https://rdar.apple.com/132458382">rdar://132458382</a>

Reviewed by Matthieu Dubet and Tim Nguyen.

Combine the existing parsing logic for both rgb() and rgba() and treat
them as aliases
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseNumericColor):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParserFastPaths.cpp: Added.
(TestWebKitAPI::TEST(CSSParserFastPaths, ParseNumericColor)):

Canonical link: <a href="https://commits.webkit.org/289053@main">https://commits.webkit.org/289053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d875388eb8907dc238db3f1b18716605dbd0853d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39559 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90302 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87242 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4970 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66225 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24039 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88199 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3770 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77349 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46533 "Found 2 new API test failures: /WPE/TestSSL:/webkit/WebKitWebView/web-socket-client-side-certificate, /WPE/TestSSL:/webkit/WebKitWebView/web-socket-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31598 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35264 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74431 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32436 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91687 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12492 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9113 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74724 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12720 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73843 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18284 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18261 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16683 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4491 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12438 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17898 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12274 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15762 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14021 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParserFastPaths.cpp:
```diff
@@ -0,0 +1,55 @@
+    };
+
+    for (auto input : testInputs)
+        EXPECT_TRUE(CSSParserFastPaths::parseSimpleColor(input));
```
👍

mdubet commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -611,18 +611,30 @@ static std::optional<SRGBA<uint8_t>> parseNumericColor(std::span<const Character
             return std::nullopt;
-        auto blue = parseColorIntOrPercentage(current, ',', expectedUnitType);
+
+        // check if blue terminates with ')' (no alpha-value provided)
```
We generally format comments like a sentence https://webkit.org/code-style-guidelines/#comments-sentences

mdubet commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -611,18 +611,30 @@ static std::optional<SRGBA<uint8_t>> parseNumericColor(std::span<const Character
+
+        // check if blue terminates with ')' (no alpha-value provided)
+        auto backup = current; // current parsing implementation consumes span, so back up in case first parse fails
+        auto blue = parseColorIntOrPercentage(current, ')', expectedUnitType);
```
It' s a bit sad that we try to parse the blue part twice in case of rgba (especially in a fast path function). 
Maybe dispatch the blue parsing based on `mightBeRGBA()` (and it would help to remove the error prone manipulation of `backup`).
Also, maybe the compiler would optimize it anyway, but using a constant to store the result of `mightBeRGB(A)`  could help.

mdubet replied:
Seems that `rgb()` and `rgba()` parse exactly the same syntax actually so we can ignore my comment about dispatch based on `mightBeRGBA()` 😬  https://drafts.csswg.org/css-color/#rgb-functions
So we should just avoid parsing twice the integer number just because of a different terminator. Using a `std::optional<char> terminator` and doing the terminator parsing manually could work.

mdubet commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -340,7 +340,7 @@ static std::optional<uint8_t> parseColorIntOrPercentage(std::span<const Characte
     skipWhile<isASCIIWhitespace>(current);
 
-    if (!skipExactly(current, terminator))
+    if (consumableTerminator && !skipExactly(current, consumableTerminator))
```
I'm surprised it's compiling without desugaring the optional ? Is it correct ? I would expect `!skipExactly(current, *consumableTerminator)`

roberto-apple replied:
Good observation. I'll double check to see if maybe it's implicitly casting to the underlying type

roberto-apple replied:
After checking, there was no implicit conversion being done.

What was actually happening is `skipExactly()` is a function template (https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/text/ParsingUtilities.h#L59-L61) so the `consumableTerminator` argument was deduced as a `std::optional<char>` for 'DelimeterType' in `skipExactly()`. Then it was used in the `==`expression and according to cppreference (https://en.cppreference.com/w/cpp/utility/optional/operator_cmp (22)), it checks to see if the std::optional has a value and if so, it then dereferences it for the comparison (in this case against a char).

In any case, I changed this call to dereference consumableTerminator to be more clear.

mdubet commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -340,7 +340,7 @@ static std::optional<uint8_t> parseColorIntOrPercentage(std::span<const Characte
     skipWhile<isASCIIWhitespace>(current);
 
-    if (!skipExactly(current, terminator))
+    if (consumableTerminator.has_value() && !skipExactly(current, *consumableTerminator))
```
Don' t need the `has_value()`, `std::optional` implements the `operator bool`
https://en.cppreference.com/w/cpp/utility/optional/operator_bool

roberto-apple replied:
Updated to use operator bool

roberto-apple committed:
CSSParserFastPaths should treat rgb() and rgba() as synonyms
https://bugs.webkit.org/show_bug.cgi?id=276761
rdar://132458382

Reviewed by Matthieu Dubet and Tim Nguyen.

Combine the existing parsing logic for both rgb() and rgba() and treat
them as aliases
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseNumericColor):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParserFastPaths.cpp: Added.
(TestWebKitAPI::TEST(CSSParserFastPaths, ParseNumericColor)):

Canonical link: https://commits.webkit.org/289053@main
